app/routes.py

```diff
@@ -1,2 +1,238 @@
+        welcome = {"project": "task list",
+                   "student": "kaitlyn",
+                   "class": "spruce"}
+        return jsonify(welcome), 200
+
+
+@tasks_bp.route("", methods=["GET"])
+def get_tasks():
+    title_from_url = request.args.get("title")
+    tasks = Task.task_arguments(title_from_url)
+
+    tasks_response = []
+    tasks_response = [task.create_dict() for task in tasks]
+
+    if not tasks_response:
+        tasks = Task.query.all()
+        tasks_response = [task.create_dict() for task in tasks]
+
+    return jsonify(tasks_response)
+
+
+@tasks_bp.route("", methods=["POST"])
+def post_tasks():
+    request_body = request.get_json()
+    if "title" not in request_body or "description" not in request_body or "completed_at" not in request_body:
+        error_dict = {"details": "Invalid data"}
+        return jsonify(error_dict), 400
+
+    new_task = Task.from_json()
+    task_response = {"task": new_task.create_dict()}
+    return jsonify(task_response), 201
+
+
+@tasks_bp.route("/<task_id>", methods=["GET"])
+def get_tasks_by_id(task_id):
+    task = Task.query.get(task_id)
+    if not task:
+        return jsonify(None), 404
+
+    task_response = {"task": task.create_dict()}
+    return jsonify(task_response), 200
+
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+def put_tasks_by_id(task_id):
+
+    task = Task.query.get(task_id)
+    if not task:
+        return jsonify(None), 404
+    form_data = request.get_json()
+
+    task.title = form_data["title"]
+    task.description = form_data["description"]
+    task.iscomplete = None
+
+    db.session.commit()
+
+    task_response = {"task": task.create_dict()}
+    return jsonify(task_response), 200
+
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+def delete_task(task_id):
+    task = Task.query.get(task_id)
+    if not task:
+        return jsonify(None), 404
+
+    db.session.delete(task)
+    db.session.commit()
+
+    delete_message = {
+        "details": f'Task {task.task_id} "{task.title}" successfully deleted'}
+
+    return jsonify(delete_message), 200
+
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def task_complete(task_id):
+    if request.method == "PATCH":
+        task = Task.query.get(task_id)
+        if not task:
+            return jsonify(None), 404
+
+        task.completed_at = datetime.datetime.now()
+        db.session.commit()
+
+        task.send_task_complete_slack_message()
+
+        task_response = {"task": task.create_dict()}
+        return jsonify(task_response), 200
+
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def task_incomplete(task_id):
+    if request.method == "PATCH":
+        task = Task.query.get(task_id)
+        if not task:
+            return jsonify(None), 404
+
+        task.completed_at = None
+        db.session.commit()
+
+        task_response = {"task": task.create_dict()}
+        return jsonify(task_response), 200
+
+
+#
+# Start Goals Routes
```

Comment:
In general, consider how the comments for Task could mostly apply to Goal as well.